ee/secureenclaverunner/secureenclaverunner.go

```diff
@@ -0,0 +1,281 @@
+//go:build darwin
+// +build darwin
+
+package secureenclaverunner
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"os/user"
+	"sync"
+	"time"
+
+	"github.com/kolide/launcher/ee/agent/types"
+	"github.com/kolide/launcher/ee/consoleuser"
+	"github.com/kolide/launcher/pkg/traces"
+)
+
+const (
+	publicEccDataKey = "publicEccData"
+)
+
+type secureEnclaveRunner struct {
+	uidPubKeyMap        map[string]*ecdsa.PublicKey
+	secureEnclaveClient secureEnclaveClient
+	store               types.GetterSetterDeleter
+	slogger             *slog.Logger
+	mux                 *sync.Mutex
+	interrupt           chan struct{}
+	interrupted         bool
+}
+
+type secureEnclaveClient interface {
+	CreateSecureEnclaveKey(uid string) (*ecdsa.PublicKey, error)
+}
+
+func New(ctx context.Context, slogger *slog.Logger, store types.GetterSetterDeleter, secureEnclaveClient secureEnclaveClient) (*secureEnclaveRunner, error) {
+	ctx, span := traces.StartSpan(ctx)
+	defer span.End()
+
+	ser := &secureEnclaveRunner{
+		uidPubKeyMap:        make(map[string]*ecdsa.PublicKey),
+		store:               store,
+		secureEnclaveClient: secureEnclaveClient,
+		slogger:             slogger.With("component", "secureenclaverunner"),
+		mux:                 &sync.Mutex{},
+		interrupt:           make(chan struct{}),
+	}
+
+	data, err := store.Get([]byte(publicEccDataKey))
+	if err != nil {
+		traces.SetError(span, fmt.Errorf("getting public ecc data from store: %w", err))
+		return nil, fmt.Errorf("getting public ecc data from store: %w", err)
+	}
+
+	if data != nil {
+		if err := json.Unmarshal(data, ser); err != nil {
+			traces.SetError(span, fmt.Errorf("unmarshaling secure enclave signer: %w", err))
+			ser.slogger.Log(ctx, slog.LevelError,
+				"unable to unmarshal secure enclave signer, data may be corrupt, wiping",
+				"err", err,
+			)
+
+			if err := store.Delete([]byte(publicEccDataKey)); err != nil {
+				traces.SetError(span, fmt.Errorf("deleting corrupt public ecc data: %w", err))
+				ser.slogger.Log(ctx, slog.LevelError,
+					"unable to unmarshal secure enclave signer, data may be corrupt, wiping",
+					"err", err,
+				)
+			}
+		}
+	}
+
+	return ser, nil
+}
+
+// Public returns the public key of the current console user
+// creating and peristing a new one if needed
+func (ser *secureEnclaveRunner) Execute() error {
+	currentRetryInterval, maxRetryInterval := 1*time.Second, 1*time.Minute
+	retryTicker := time.NewTicker(currentRetryInterval)
+	defer retryTicker.Stop()
+
+	for {
+		if _, err := ser.currentConsoleUserKey(context.TODO()); err != nil {
+			ser.slogger.Log(context.TODO(), slog.LevelError,
+				"getting current console user key, will retry",
+				"err", err,
+			)
+
+			if currentRetryInterval < maxRetryInterval {
```

Comment:
lazy backoff ... is there a better way or is this good enough?

Reply:
actaully, I don't this is backing off, need to reset the ticker

Reply:
I like this and think it's good enough! Since you use it here and in the tpmrunner (and since I can imagine it being useful in other areas too) I wonder if we might want to extract it into `pkg/backoff`?

Reply:
Oh, I just saw we do actually have a `Backoff` in that package -- I forgot we had anything in there besides `WaitFor` 😅 . Maybe an opportunity to use that/combine?

Reply:
I didn't really like what was in there for this application, because there was no way to stop, so I created a custom ticker with "Multiplicative Backoff" (duration * ticks) ... maybe went a little overboard =)